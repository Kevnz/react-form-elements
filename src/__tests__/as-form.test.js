import React, { useRef, useEffect } from 'react'
import { render, fireEvent, cleanup } from 'react-testing-library'

import {
  TextBox,
  Checkbox,
  Range,
  DateTime,
  Telephone,
  DropDown,
  Option,
  OptionGroup,
} from '../index'

const Form = ({ name, onSubmit, children }) => {
  return (
    <form name={name} onSubmit={onSubmit}>
      {' '}
      {children}
    </form>
  )
}

const App = () => {
  const references = {
    textBox: useRef(null),
    checkBox: useRef(null),
    range: useRef(null),
    date: useRef(null),
    telephone: useRef(null),
    dropdown: useRef(null),
  }

  const getFormValues = () => {
    if (references.textBox.current === null)
      return {
        textBox: '',
        checkBox: '',
        range: '',
        date: '',
        telephone: '',
        dropdown: '',
      }
    return {
      textBox: references.textBox.current.getValue(),
      checkBox: references.checkBox.current.getValue(),
      range: references.range.current.getValue(),
      date: references.date.current.getValue(),
      telephone: references.telephone.current.getValue(),
      dropdown: references.dropdown.current.getValue(),
    }
  }
  let formValues = getFormValues()

  useEffect(() => {
    // Update the document title using the browser API
    formValues = getFormValues()
  })

  return (
    <div>
      <h1>React Form Elements</h1>
      <Form
        name="testForm"
        onSubmit={e => {
          e.target.preventDefault()
        }}
      >
        <TextBox label="My Label" ref={references.textBox} />
        <Checkbox
          label="My Checkbox"
          ref={references.checkBox}
          initialValue="{true}"
        />
        <DateTime label="My Date" ref={references.date} />
        <Range label="My Range" ref={references.range} />
        <Telephone label="Telephone" ref={references.telephone} />
        <DateTime label="My DateTime" type="datetime" />
        <DateTime label="My Month" type="month" />
        <DateTime label="My Week" type="week" />
        <DateTime label="My DateTime" type="datetime-local" />
        <DateTime label="My Time" type="time" />
        <DropDown
          label="My Drop Down"
          initialValue="2"
          data-testid="dd1"
          ref={references.dropdown}
        >
          <OptionGroup label="First Group">
            <Option initialValue="1">First</Option>
            <Option initialValue="2">Second</Option>
            <Option initialValue="3" label="Third" />
          </OptionGroup>
          <OptionGroup label="Second Group">
            <Option initialValue="11">Second First</Option>
            <Option initialValue="12">Second Second</Option>
            <Option initialValue="13" label="Second Third" />
          </OptionGroup>
          <OptionGroup label="Third Group">
            <Option initialValue="21">Third First</Option>
            <Option initialValue="22">Third Second</Option>
            <Option initialValue="23" label="Third Third" />
          </OptionGroup>
        </DropDown>
        <button
          onClick={e => {
            e.preventDefault()
          }}
        >
          Check
        </button>
      </Form>
      <div data-testid="ref-out">
        <ul>
          <li data-testid="textbox">TextBox: {formValues.textBox}</li>
          <li data-testid="checkbox">Checkbox: {formValues.checkBox}</li>
          <li data-testid="range">Range: {formValues.range}</li>
          <li data-testid="date">Date: {formValues.date}</li>
          <li data-testid="telephone">Telephone: {formValues.telephone}</li>
        </ul>
      </div>
    </div>
  )
}

describe('The form components as a form', () => {
  afterEach(cleanup)
  it('should render and change accordingly', () => {
    const { container, getByLabelText, getByTestId, rerender } = render(<App />)
    const textBox = getByLabelText('My Label')
    const dd = getByTestId('dd1')
    expect(container.firstChild).toMatchSnapshot()
    const dropDownValue = dd.value
    expect(dropDownValue).toBe('2')
    fireEvent.change(textBox, { target: { value: 'Changed Value' } })
    fireEvent.change(dd, { target: { value: '22' } })
    const changedDropDownValue = dd.value
    expect(changedDropDownValue).toBe('22')
    rerender(<App />)
    expect(textBox.value).toBe('Changed Value')
    expect(container.firstChild).toMatchSnapshot()
  })
})
