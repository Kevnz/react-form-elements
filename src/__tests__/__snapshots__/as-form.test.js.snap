// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The form components as a form should render properly 1`] = `
<div>
  <h1>
    React Form Elements
  </h1>
  <form>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-1-1"
      >
        My Label
      </label>
      <input
        id="id-1-1"
        name="TextBox"
        type="text"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-2-1"
      >
        My Checkbox
      </label>
      <input
        checked=""
        id="id-2-1"
        type="checkbox"
        value="as-test-check"
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-3-1"
      >
        My Date
      </label>
      <input
        id="id-3-1"
        name="TextBox"
        type="date"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-4-1"
      >
        My Range
      </label>
      <input
        id="id-4-1"
        max="100"
        min="0"
        step="1"
        type="range"
        value="50"
      />
      <span>
        50
      </span>
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-5-1"
      >
        Telephone
      </label>
      <input
        id="id-5-1"
        name="TextBox"
        type="tel"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-6-1"
      >
        My DateTime
      </label>
      <input
        id="id-6-1"
        name="TextBox"
        type="datetime"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-7-1"
      >
        My Month
      </label>
      <input
        id="id-7-1"
        name="TextBox"
        type="month"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-8-1"
      >
        My Week
      </label>
      <input
        id="id-8-1"
        name="TextBox"
        type="week"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-9-1"
      >
        My DateTime
      </label>
      <input
        id="id-9-1"
        name="TextBox"
        type="datetime-local"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-10-1"
      >
        My Time
      </label>
      <input
        id="id-10-1"
        name="TextBox"
        type="time"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-11-1"
      >
        My Drop Down
      </label>
      <select
        data-testid="dd1"
        id="id-11-1"
      >
        <optgroup
          label="First Group"
        >
          <option
            value="1"
          >
            First
          </option>
          <option
            value="2"
          >
            Second
          </option>
          <option
            value="3"
          >
            Third
          </option>
        </optgroup>
        <optgroup
          label="Second Group"
        >
          <option
            value="11"
          >
            Second First
          </option>
          <option
            value="12"
          >
            Second Second
          </option>
          <option
            value="13"
          >
            Second Third
          </option>
        </optgroup>
        <optgroup
          label="Third Group"
        >
          <option
            value="21"
          >
            Third First
          </option>
          <option
            value="22"
          >
            Third Second
          </option>
          <option
            value="23"
          >
            Third Third
          </option>
        </optgroup>
      </select>
    </div>
    <button>
      Check
    </button>
  </form>
  <div
    data-testid="ref-out"
  >
    <ul>
      <li
        data-testid="textbox"
      >
        TextBox: 
        
      </li>
      <li
        data-testid="checkbox"
      >
        Checkbox: 
        
      </li>
      <li
        data-testid="range"
      >
        Range: 
        
      </li>
      <li
        data-testid="date"
      >
        Date: 
        
      </li>
      <li
        data-testid="telephone"
      >
        Telephone: 
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`The form components as a form should render properly 2`] = `
<div>
  <h1>
    React Form Elements
  </h1>
  <form>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-14-1"
      >
        My Label
      </label>
      <input
        id="id-14-1"
        name="TextBox"
        type="text"
        value="Changed Value"
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-15-1"
      >
        My Checkbox
      </label>
      <input
        checked=""
        id="id-15-1"
        type="checkbox"
        value="as-test-check"
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-16-1"
      >
        My Date
      </label>
      <input
        id="id-16-1"
        name="TextBox"
        type="date"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-17-1"
      >
        My Range
      </label>
      <input
        id="id-17-1"
        max="100"
        min="0"
        step="1"
        type="range"
        value="50"
      />
      <span>
        50
      </span>
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-18-1"
      >
        Telephone
      </label>
      <input
        id="id-18-1"
        name="TextBox"
        type="tel"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-19-1"
      >
        My DateTime
      </label>
      <input
        id="id-19-1"
        name="TextBox"
        type="datetime"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-20-1"
      >
        My Month
      </label>
      <input
        id="id-20-1"
        name="TextBox"
        type="month"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-21-1"
      >
        My Week
      </label>
      <input
        id="id-21-1"
        name="TextBox"
        type="week"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-22-1"
      >
        My DateTime
      </label>
      <input
        id="id-22-1"
        name="TextBox"
        type="datetime-local"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-23-1"
      >
        My Time
      </label>
      <input
        id="id-23-1"
        name="TextBox"
        type="time"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-24-1"
      >
        My Drop Down
      </label>
      <select
        data-testid="dd1"
        id="id-24-1"
      >
        <optgroup
          label="First Group"
        >
          <option
            value="1"
          >
            First
          </option>
          <option
            value="2"
          >
            Second
          </option>
          <option
            value="3"
          >
            Third
          </option>
        </optgroup>
        <optgroup
          label="Second Group"
        >
          <option
            value="11"
          >
            Second First
          </option>
          <option
            value="12"
          >
            Second Second
          </option>
          <option
            value="13"
          >
            Second Third
          </option>
        </optgroup>
        <optgroup
          label="Third Group"
        >
          <option
            value="21"
          >
            Third First
          </option>
          <option
            value="22"
          >
            Third Second
          </option>
          <option
            value="23"
          >
            Third Third
          </option>
        </optgroup>
      </select>
    </div>
    <button>
      Check
    </button>
  </form>
  <div
    data-testid="ref-out"
  >
    <ul>
      <li
        data-testid="textbox"
      >
        TextBox: 
        Changed Value
      </li>
      <li
        data-testid="checkbox"
      >
        Checkbox: 
        as-test-check
      </li>
      <li
        data-testid="range"
      >
        Range: 
        50
      </li>
      <li
        data-testid="date"
      >
        Date: 
        
      </li>
      <li
        data-testid="telephone"
      >
        Telephone: 
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`The form components as a form should render properly c2 1`] = `
<div>
  <h1>
    React Form Elements
  </h1>
  <form>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-38-1"
      >
        My Label
      </label>
      <input
        id="id-38-1"
        name="TextBox"
        type="text"
        value="Changed Value"
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-39-1"
      >
        My Checkbox
      </label>
      <input
        checked=""
        id="id-39-1"
        type="checkbox"
        value="as-test-check"
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-40-1"
      >
        My Date
      </label>
      <input
        id="id-40-1"
        name="TextBox"
        type="date"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-41-1"
      >
        My Range
      </label>
      <input
        id="id-41-1"
        max="100"
        min="0"
        step="1"
        type="range"
        value="50"
      />
      <span>
        50
      </span>
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-42-1"
      >
        Telephone
      </label>
      <input
        id="id-42-1"
        name="TextBox"
        type="tel"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-43-1"
      >
        My DateTime
      </label>
      <input
        id="id-43-1"
        name="TextBox"
        type="datetime"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-44-1"
      >
        My Month
      </label>
      <input
        id="id-44-1"
        name="TextBox"
        type="month"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-45-1"
      >
        My Week
      </label>
      <input
        id="id-45-1"
        name="TextBox"
        type="week"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-46-1"
      >
        My DateTime
      </label>
      <input
        id="id-46-1"
        name="TextBox"
        type="datetime-local"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-47-1"
      >
        My Time
      </label>
      <input
        id="id-47-1"
        name="TextBox"
        type="time"
        value=""
      />
    </div>
    <div
      class="rfe-form__row"
    >
      <label
        for="id-48-1"
      >
        My Drop Down
      </label>
      <select
        data-testid="dd1"
        id="id-48-1"
      >
        <optgroup
          label="First Group"
        >
          <option
            value="1"
          >
            First
          </option>
          <option
            value="2"
          >
            Second
          </option>
          <option
            value="3"
          >
            Third
          </option>
        </optgroup>
        <optgroup
          label="Second Group"
        >
          <option
            value="11"
          >
            Second First
          </option>
          <option
            value="12"
          >
            Second Second
          </option>
          <option
            value="13"
          >
            Second Third
          </option>
        </optgroup>
        <optgroup
          label="Third Group"
        >
          <option
            value="21"
          >
            Third First
          </option>
          <option
            value="22"
          >
            Third Second
          </option>
          <option
            value="23"
          >
            Third Third
          </option>
        </optgroup>
      </select>
    </div>
    <button>
      Check
    </button>
  </form>
  <div
    data-testid="ref-out"
  >
    <ul>
      <li
        data-testid="textbox"
      >
        TextBox: 
        Changed Value
      </li>
      <li
        data-testid="checkbox"
      >
        Checkbox: 
        as-test-check
      </li>
      <li
        data-testid="range"
      >
        Range: 
        50
      </li>
      <li
        data-testid="date"
      >
        Date: 
        
      </li>
      <li
        data-testid="telephone"
      >
        Telephone: 
        
      </li>
    </ul>
  </div>
</div>
`;
